language-support/java/bindings/src/test/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReaderTest.java

```diff
@@ -0,0 +1,388 @@
+            timestampUTC(9999, Month.DECEMBER, 31, 23, 59, 59, 999999)),
+        //            eq("\"1990-11-09T04:30:23.1234569Z\"", timestampUTC(1990, Month.NOVEMBER,
+        // 9,  4, 30, 23, 123457)), // TODO(raphael-speyer-da): Appropriate rounding
+        eq("\"1990-11-09T04:30:23Z\"", timestampUTC(1990, Month.NOVEMBER, 9, 4, 30, 23, 0)),
+        eq(
+            "\"1990-11-09T04:30:23.123Z\"",
+            timestampUTC(1990, Month.NOVEMBER, 9, 4, 30, 23, 123000)),
+        eq("\"0001-01-01T00:00:00Z\"", timestampUTC(1, Month.JANUARY, 1, 0, 0, 0, 0)));
+  }
+
+  @Test
+  void testDate() throws IOException {
+    checkReadAll(
+        JsonLfReader::date,
+        eq("\"2019-06-18\"", date(2019, Month.JUNE, 18)),
+        eq("\"9999-12-31\"", date(9999, Month.DECEMBER, 31)),
+        eq("\"0001-01-01\"", date(1, Month.JANUARY, 1)));
+  }
+
+  @Test
+  void testParty() throws IOException {
+    checkReadAll(JsonLfReader::party, eq("\"Alice\"", "Alice"));
+  }
+
+  @Test
+  void testText() throws IOException {
+    checkReadAll(JsonLfReader::text, eq("\"\"", ""), eq("\" \"", " "), eq("\"hello\"", "hello"));
+  }
+
+  @Test
+  void testContractId() throws IOException {
+    checkReadAll(r -> r.contractId(Tmpl.Cid::new), eq("\"deadbeef\"", new Tmpl.Cid("deadbeef")));
+  }
+
+  @Test
+  void testEnum() throws IOException {
+    checkReadAll(
+        r -> r.enumeration(Suit.class),
+        eq("\"Hearts\"", Suit.Hearts),
+        eq("\"Diamonds\"", Suit.Diamonds),
+        eq("\"Clubs\"", Suit.Clubs),
+        eq("\"Spades\"", Suit.Spades));
+  }
+
+  @Test
+  void testList() throws IOException {
+    checkReadAll(r -> r.list(r.int64()), eq("[]", emptyList()), eq("[1,2]", asList(1L, 2L)));
+  }
+
+  @Test
+  void testTextMap() throws IOException {
+    checkReadAll(
+        r -> r.textMap(r.int64()),
+        eq("{}", emptyMap()),
+        eq("{\"foo\":1, \"bar\": 2}", java.util.Map.of("foo", 1L, "bar", 2L)));
+  }
+
+  @Test
+  void testGenMap() throws IOException {
+    checkReadAll(
+        r -> r.genMap(r.text(), r.int64()),
+        eq("[]", emptyMap()),
+        eq("[[\"foo\", 1], [\"bar\", 2]]", java.util.Map.of("foo", 1L, "bar", 2L)));
+  }
+
+  @Test
+  void testOptionalNonNested() throws IOException {
+    checkReadAll(
+        r -> r.optional(r.int64()), eq("null", Optional.empty()), eq("42", Optional.of(42L)));
+  }
+
+  @Test
+  void testOptionalNested() throws IOException {
+    checkReadAll(
+        r -> r.optionalNested(r.optional(r.int64())),
+        eq("null", Optional.empty()),
+        eq("[]", Optional.of(Optional.empty())),
+        eq("[42]", Optional.of(Optional.of(42L))));
+  }
+
+  @Test
+  void testOptionalNestedDeeper() throws IOException {
+    checkReadAll(
+        r -> r.optionalNested(r.optionalNested(r.optional(r.int64()))),
+        eq("null", Optional.empty()),
+        eq("[]", Optional.of(Optional.empty())),
+        eq("[[]]", Optional.of(Optional.of(Optional.empty()))),
```

Comment:
Unit tests look really comprehensive. Maybe add one for this nested list case when fields are non-optional?

Reply:
I added one for `Optional[List[List[Int64]]]`